register.go

```diff
@@ -127,33 +235,51 @@ func unregister() error {
 		return fmt.Errorf("warning: the system is already unregistered")
 	}
 
-	if err := conn.Object("com.redhat.RHSM1", "/com/redhat/RHSM1/Unregister").Call("com.redhat.RHSM1.Unregister.Unregister", dbus.Flags(0), map[string]string{}, "").Err; err != nil {
-		return unpackError(err)
+	if err := conn.Object(
+		"com.redhat.RHSM1",
+		"/com/redhat/RHSM1/Unregister").Call(
+			"com.redhat.RHSM1.Unregister.Unregister",
+			dbus.Flags(0),
+			map[string]string{},
+			"").Err; err != nil {
+		return unpackErrorMsg(err)
 	}
 
 	return nil
 }
 
-func unpackError(err error) error {
+// RHSMError is used for parsing JSON document returned by D-Bus methods.
+type RHSMError struct {
+	Exception string `json:"exception"`
+	Severity  string `json:"severity"`
+	Message   string `json:"message"`
+}
+
+// unpackError tries to unpack JSON document (part of error) into the structure RHSMError. When it is
+// not possible to parse error into structure, then return empty structure and error. When it is
+// possible to parse error into structure, then returned structure contains valid data and err is nil.
+func unpackError(err error) (RHSMError, error) {
```

Comment:
There's a convention I've discovered that can ensure your caller doesn't mistakenly attempt to use the zero value of the struct returned by your function. If you return a pointer to `RHSMError`, you can return `nil` when returning an error. This prevents a caller of your function from ignoring the error (intentionally or unintentionally) and proceeding as if the function returned successfully. Returning a zero value struct along with an error can lead to silent problems later on when a struct field is accessed (and returns *its* zero value). Returning `nil` will cause that field access to panic and abort the program.

Reply:
Yeah, I like this pattern. I will change the code.